File: briar-android/src/main/java/org/briarproject/briar/android/contact/TelegramInboxThreadViewHolder.java
package org.briarproject.briar.android.contact;

import android.content.Context;
import android.content.res.Resources;
import android.view.View;
import android.widget.TextView;

import org.briarproject.briar.R;
import org.briarproject.nullsafety.NotNullByDefault;

import androidx.annotation.UiThread;

import static org.briarproject.briar.android.util.UiUtils.formatDate;

@UiThread
@NotNullByDefault
final class TelegramInboxThreadViewHolder extends
		androidx.recyclerview.widget.RecyclerView.ViewHolder {

	private final TextView title;
	private final TextView date;
	private final TextView preview;

	TelegramInboxThreadViewHolder(View view) {
		super(view);
		title = view.findViewById(R.id.telegramThreadTitle);
		date = view.findViewById(R.id.telegramThreadDate);
		preview = view.findViewById(R.id.telegramThreadPreview);
	}

	void bind(TelegramInboxThreadItem item,
			OnContactClickListener<TelegramInboxThreadItem> listener) {
		title.setText(titleText(title.getResources(), item));
		date.setVisibility(dateVisibility(item));
		date.setText(dateText(date.getContext(), item));
		if (item.isPreviewLoading()) {
			preview.setText(R.string.telegram_thread_preview_loading);
		} else if (item.hasPreviewText()) {
			preview.setText(previewText(preview.getResources(), item));
		} else {
			preview.setText(R.string.telegram_thread_preview_empty);
		}
		itemView.setOnClickListener(view -> listener.onItemClick(view, item));
	}

	static CharSequence titleText(Resources resources,
			TelegramInboxThreadItem item) {
		String title = item.getTitle();
		if (title.trim().isEmpty()) {
			return resources.getString(R.string.telegram_thread_title_fallback);
		}
		return title;
	}

	static int dateVisibility(TelegramInboxThreadItem item) {
		return item.getLatestActivityMillis() > 0 ? View.VISIBLE : View.GONE;
	}

	static CharSequence dateText(Context context, TelegramInboxThreadItem item) {
		long latestActivityMillis = item.getLatestActivityMillis();
		if (latestActivityMillis <= 0) return "";
		return formatDate(context, latestActivityMillis);
	}

	static CharSequence previewText(Resources resources,
			TelegramInboxThreadItem item) {
		if (!item.isLastMessageOutgoing()) return item.getPreviewText();
		return resources.getString(R.string.telegram_thread_preview_outgoing,
				item.getPreviewText());
	}
}

File: briar-android/src/main/kotlin/org/briarproject/briar/android/contact/TelegramInboxThreadViewHolder.kt
package org.briarproject.briar.android.contact

import android.content.Context
import android.content.res.Resources
import android.view.View
import android.widget.ImageView
import android.widget.TextView
import androidx.annotation.UiThread
import androidx.recyclerview.widget.RecyclerView
import org.briarproject.briar.R
import org.briarproject.briar.android.util.UiUtils.formatDate

@UiThread
class TelegramInboxThreadViewHolder(view: View) : RecyclerView.ViewHolder(view) {
	private val sourceIcon: ImageView = view.findViewById(R.id.telegramThreadIcon)
	private val title: TextView = view.findViewById(R.id.telegramThreadTitle)
	private val date: TextView = view.findViewById(R.id.telegramThreadDate)
	private val badge: TextView = view.findViewById(R.id.telegramThreadBadge)
	private val preview: TextView = view.findViewById(R.id.telegramThreadPreview)

	fun bind(item: ConnectorInboxThreadItem,
		listener: OnContactClickListener<ConnectorInboxThreadItem>) {
		title.text = titleText(title.resources, item)
		date.visibility = dateVisibility(item)
		date.text = dateText(date.context, item)
		sourceIcon.contentDescription =
			sourceContentDescription(sourceIcon.resources, item)
		badge.text = sourceLabel(item)
		preview.text = when {
			item.isPreviewLoading ->
				preview.resources.getString(R.string.telegram_thread_preview_loading)
			item.hasPreviewText() -> previewText(preview.resources, item)
			else -> preview.resources.getString(R.string.telegram_thread_preview_empty)
		}
		itemView.setOnClickListener { view -> listener.onItemClick(view, item) }
	}

	companion object {
		@JvmStatic
		fun titleText(resources: Resources,
			item: ConnectorInboxThreadItem): CharSequence {
			val title = item.title
			return if (title.trim().isEmpty())
				resources.getString(R.string.telegram_thread_title_fallback)
			else title
		}

		@JvmStatic
		fun sourceLabel(item: ConnectorInboxThreadItem): CharSequence =
			item.connectorSource.displayName

		@JvmStatic
		fun sourceContentDescription(resources: Resources,
			item: ConnectorInboxThreadItem): CharSequence =
			resources.getString(
				R.string.connector_thread_source_content_description,
				item.connectorSource.displayName
			)

		@JvmStatic
		fun dateVisibility(item: ConnectorInboxThreadItem): Int =
			if (item.latestActivityMillis > 0) View.VISIBLE else View.GONE

		@JvmStatic
		fun dateText(context: Context, item: ConnectorInboxThreadItem): CharSequence {
			val latestActivityMillis = item.latestActivityMillis
			if (latestActivityMillis <= 0) return ""
			return formatDate(context, latestActivityMillis)
		}

		@JvmStatic
		fun previewText(resources: Resources,
			item: ConnectorInboxThreadItem): CharSequence =
			if (!item.isLastMessageOutgoing) item.previewText
			else resources.getString(R.string.telegram_thread_preview_outgoing,
				item.previewText)
	}
}
